### PR TITLE
Hide file remove link for autogenerated _product:* files

### LINKED ISSUE #3777: [production][package#remove_file] <status code="403">
  <summary>package '_product:openSUSE-dvd5-dvd-i586' is read-only if a '_prod...
package#remove_file threw this 0 times
<pre>
&lt;status code=&quot;403&quot;&gt;
  &lt;summary&gt;package &#39;_product:openSUSE-dvd5-dvd-i586&#39; is read-only if a &#39;_product&#39; package exists&lt;/summary&gt;
  &lt;details&gt;403 package &#39;_product:openSUSE-dvd5-dvd-i586&#39; is read-only if a &#39;_product&#39; package exists&lt;/details&gt;
&lt;/status&gt;

</pre>
[https://build.opensuse.org/package/remove_file/home:aplanas:branches:openSUSE:Factory/_product:openSUSE-dvd5-dvd-i586?filename=openSUSE-dvd5-dvd-i586.kiwi](https://build.opensuse.org/package/remove_file/home:aplanas:branches:openSUSE:Factory/_product:openSUSE-dvd5-dvd-i586?filename=openSUSE-dvd5-dvd-i586.kiwi)"

[See this exception on Errbit](https://errbit-opensuse.herokuapp.com/apps/5620ca2bdc71fa00b1000000/problems/59b003954684900006e2c344)

### PR DESCRIPTION
**Pull request by bgeuken**
_product packages contain instructions that automatically created
additional _product:* packages. These packages contain files that were
showing a delete file icon.
Though all these files are auto-generated as well and attempting to
delete them results in an error ("package '_product:foo' is read-only".

Fixes #3777

### CODE REVIEW

**Comment by bgeuken** [comment]
I am going to add a test for this -> added 'do not merge' label

File: src/api/spec/helpers/webui/package_helper_spec.rb [hunk 1]
```diff
@@ -2,6 +2,14 @@
 require 'rantly/rspec_extensions'
 
 RSpec.describe Webui::PackageHelper, type: :helper do
+  describe '#removable_file?' do
+    it { expect(removable_file?(file_name: 'foo',               package_name: 'bar')).to                  be true }
+    it { expect(removable_file?(file_name: '_service',          package_name: 'bar')).to                  be true }
+    it { expect(removable_file?(file_name: '_service:sub_file', package_name: 'bar')).to                  be false }
```
Comment (hennevogel) [inline src/api/spec/helpers/webui/package_helper_spec.rb]: What happens to the `_service:sub_file` if I remove the _service file?

File: src/api/app/helpers/webui/package_helper.rb [hunk 2]
```diff
@@ -1,4 +1,8 @@
 module Webui::PackageHelper
+  def removable_file?(file_name:, package_name:)
```
Comment (ChrisBr) [inline src/api/app/helpers/webui/package_helper.rb]: Why named parameters?

File: src/api/.rubocop.yml [hunk 3]
```diff
@@ -118,6 +118,7 @@ Style/MethodCallWithArgsParentheses:
     - assert_select
     - assert_xml_tag
     - authorize
+    - be
```
Comment (ChrisBr) [inline src/api/.rubocop.yml]: Not sure if this belongs to the same commit

File: src/api/spec/helpers/webui/package_helper_spec.rb (hunk 1 above)
Comment (hennevogel) [reply src/api/spec/helpers/webui/package_helper_spec.rb]: ok

**Comment by bgeuken** [comment]
> ChrisBr requested changes 12 minutes ago
> Do I understand right that in an automatically created package with the name _product I can not delete any file?

You can: https://github.com/openSUSE/open-build-service/pull/5649/files#diff-1dc755da48c69678488f8687effa3a18R9

The `_product` package is created by a human (or bot). From that OBS creates a couple of `_product:*` files. And these files can not be removed.

**Comment by bgeuken** [comment]
[This](https://build.opensuse.org/project/show/home:varkoly:OSS-4-0:stable) would be an example for such a _product file

File: src/api/app/helpers/webui/package_helper.rb (hunk 2 above)
Comment (bgeuken) [reply src/api/app/helpers/webui/package_helper.rb]: Because `removable_file?('something', 'something_else')` is harder to get than removable_file?(file_name: 'something', package_name: 'something_else').

This could of course also be achieved with `def removable_file?(options)`, but that would require to write:

```ruby
 !option[:file_name].start_with?('_service:') && !option[:package_name].start_with?('_product:')
```

within the method.

File: src/api/.rubocop.yml (hunk 3 above)
Comment (bgeuken) [reply src/api/.rubocop.yml]: I moved it to another commit

File: src/api/app/helpers/webui/package_helper.rb (hunk 2 above)
Comment (ChrisBr) [reply src/api/app/helpers/webui/package_helper.rb]: Well, this is how function parameters work and we do every else? Options make sense if there are options or optional fields. In this case, both parameters are required always!

**Comment by ChrisBr** [comment]
> This would be an example for such a _product file

can you link to a _product:* file? I wasn't able to find one in the provided project...

**Comment by ChrisBr** [comment]
> can you link to a _product:* file? I wasn't able to find one in the provided project...

Got it now, sorry! You can not delete any file in a ``_product:`` package

File: src/api/app/helpers/webui/package_helper.rb (hunk 2 above)
Comment (bgeuken) [reply src/api/app/helpers/webui/package_helper.rb]: > In this case, both parameters are required always!

Exactly! That's why I sticked with [keyword arguments](https://robots.thoughtbot.com/ruby-2-keyword-arguments#required-keyword-arguments). And for the other reasons:wink:

**Comment by bgeuken** [comment]
> ChrisBr commented
> This would be an example for such a _product filecan you link to a _product:* file? I wasn't able to find one in the provided project...


Yes, check out the one from [here](https://github.com/openSUSE/open-build-service/pull/5649#issuecomment-413148974)

File: src/api/app/helpers/webui/package_helper.rb (hunk 2 above)
Comment (ChrisBr) [reply src/api/app/helpers/webui/package_helper.rb]: I still don't get it! We use everywhere else positional arguments if it is a required parameter so I don't really see the point here to do it different, sorry! You always have to use this function now with specifying keyword arguments.

Comment (bgeuken) [reply src/api/app/helpers/webui/package_helper.rb]: > You always have to use this function now with specifying keyword arguments.

Huh? The keyword arguments are specified in the method definition. The method call is done by passing a hash with those two values.
I don't see how this is different compared other method calls.

And this is plain Ruby. I don't see any reason not to use it. And I am reluctant to start distinguishing between Ruby syntax we want to use and Ruby syntax we don't want to use.

It's normal that we don't have occasions of Ruby features that have been introduced over time and we couldn't use because one of the Ruby versions we support in OBS would not support $ruby_feature.
This potentially keeps us from using new Ruby features / syntax changes / etc. which I would find sad.

Regardless of that, if we don't want to use this (or other idioms of Ruby like [named captures](https://ruby-doc.org/core-2.5.0/Regexp.html#method-i-named_captures) to name one) in future we need to prohibit somehow, eg. adding a Rubocop rule for this.

@openSUSE/obs-frontend opinions?

Comment (bgeuken) [reply src/api/app/helpers/webui/package_helper.rb]: btw we have a few places already that uses keyword arguments:
* [bs_request.rb](https://github.com/openSUSE/open-build-service/blob/master/src/api/app/models/bs_request.rb#L1210)
* [project.rb](https://github.com/openSUSE/open-build-service/blob/master/src/api/app/models/project.rb#L841)

**Comment by adrianschroeter** [comment]
please note that you can delete/modify files in _product:* if _product is not there (eg in release target projects).

File: src/api/app/helpers/webui/package_helper.rb (hunk 2 above)
Comment (ChrisBr) [reply src/api/app/helpers/webui/package_helper.rb]: Two uses vs. 1000 :smile: 

I totally agree that this is a neat feature in Ruby. But we don't need to adapt every new feature Ruby does introduce. Positional parameters are still the standard way in ruby to define parameters in function calls. I would like to avoid to use different syntax everywhere to basically achieve the same.

Comment (ChrisBr) [reply src/api/app/helpers/webui/package_helper.rb]: Of course it **can** improve readability. However, if I call a function with cryptic variables then something before already went wrong. E.g.:

```
removable_file?(a, b) # no idea 

vs. 

removable_file?(file[:name], package_name: @package.name) # I know what these variables are
```

So in this case, I don't think it improves readability a lot and I also think it not necessary and we should go with the **good old** positional parameters.

And with keyword arguments you also need to know exactly the name of the parameters, so if you change it, you also need to change all callers (of course also the case when you change the order of positional arguments, however, I claim that this does not happen that often).

tl;dr: positional arguments are still the standard way to pass parameters to methods in Ruby and most other programming language. I would stick to the standard way except there is a reason to not to do. In this case, with two straight forward parameters, I don't see the reason to introduce this non common syntax.

Comment (hennevogel) [reply src/api/app/helpers/webui/package_helper.rb]: This is basically a style question and a nitpick. I have nothing against using keyword arguments. The difference in readability is tiny and everybody should have keyword arguments in their toolbelt.

Comment (vpereira) [reply src/api/app/helpers/webui/package_helper.rb]: hm the difference is that after ruby 2.1, if you don't pass any argument you get an error like:

```
def foo(bar:)
  puts bar
end

foo # => ArgumentError: missing keyword: bar
foo(bar: 'baz') # => 'baz'
```

Comment (ChrisBr) [reply src/api/app/helpers/webui/package_helper.rb]: Before ruby 2.1 there were no keyword arguments.

Comment (ChrisBr) [reply src/api/app/helpers/webui/package_helper.rb]: @vpereira can you elaborate why you click :-1: ?

Comment (ChrisBr) [reply src/api/app/helpers/webui/package_helper.rb]: > hm the difference is that after ruby 2.1, if you don't pass any argument you get an error like:

The very same thing happens with good old positional arguments!

Comment (ChrisBr) [reply src/api/app/helpers/webui/package_helper.rb]: > This is basically a style question and a nitpick. 

If this is the only issue in this PR, I'm fine with merging it.

The style question we should resolve as I don't want to end up with half of methods called with positional arguments and the other half with keyword arguments. I propose:

* default: positional arguments
* use keyword arguments for optional arguments

This is like most ruby programs do it!

Comment (vpereira) [reply src/api/app/helpers/webui/package_helper.rb]: Sorry, just that I think it isn't just about readability but correctness and safety. You can't pass nil as parameter

Comment (ChrisBr) [reply src/api/app/helpers/webui/package_helper.rb]: Of course you can:

```
def foo(a:)
  puts a
end

foo(a: nil)
```

This buys you nothing ...

Comment (vpereira) [reply src/api/app/helpers/webui/package_helper.rb]: Ya but not

   foo nil

Comment (ChrisBr) [reply src/api/app/helpers/webui/package_helper.rb]: Sure not but I don't get why this is safer or correcter? This does not prevent you from passing wrong parameters at all!

Does nobody agree that positional arguments are still the standard way to define a method signature in Ruby?

Comment (eduardoj) [reply src/api/app/helpers/webui/package_helper.rb]: @bgeuken:

> @openSUSE/obs-frontend opinions?

I prefer the required keyword arguments approach.

@ChrisBr:

> Does nobody agree that positional arguments are still the standard way to define a method signature in Ruby?

I agree. But in this case I need to see for what are they going to be used each one of the parameters. And only the position of the parameters in the call doesn't give me that information.

File: src/api/spec/helpers/webui/package_helper_spec.rb (hunk 1 above)
Comment (bgeuken) [reply src/api/spec/helpers/webui/package_helper_spec.rb]: Yes, they do. Same for the _product:* files

File: src/api/app/helpers/webui/package_helper.rb (hunk 2 above)
Comment (dmarcoux) [reply src/api/app/helpers/webui/package_helper.rb]: I don't mind, I find both fine.

**Comment by bgeuken** [comment]
Added a check for the existence of the main _product file

**Comment by bgeuken** [comment]
I extended the code, and tests, to cover _product:* files that were not created by a _product file.

Please have a second look. @adrianschroeter and Co

**Comment by bgeuken** [comment]
@openSUSE/obs-frontend Please have another look

File: src/api/app/models/package.rb [hunk 4]
```diff
@@ -804,7 +810,7 @@ def delete_on_backend
 
     # not really packages...
     # everything below _product:
-    return if name =~ /\A_product:\w[-+\w\.]*\z/ && master_product_object != self
+    return if belongs_to_product? && project.packages.product.exists?
```
Comment (eduardoj) [inline src/api/app/models/package.rb]: `&& project.packages.product.exists?` is already included in `belongs_to_product?` method. No need to be included here, again.

File: src/api/spec/models/package_spec.rb [hunk 5]
```diff
@@ -731,5 +731,29 @@
       it { expect(package.ignored_requests).to be_nil }
     end
   end
+
+  describe '#belongs_to_product?' do
+    context 'a product package (_product)' do
+      subject { create(:package, name: '_product') }
```
Comment (dmarcoux) [inline src/api/spec/models/package_spec.rb]: [Nitpick] I think the subject could be `create(:package, name: '_product').belongs_to_product?`, then the following examples would be `it { is_expected ... }`

Comment (bgeuken) [reply src/api/spec/models/package_spec.rb]: I actuallyfind it a bit clearer as it is right now

**Comment by hennevogel** [comment]
@adrianschroeter can you have a look again please?

File: src/api/app/models/package.rb [hunk 6]
```diff
@@ -357,8 +359,12 @@ def kiwi_image_outdated?
 
   def master_product_object
     # test _product permissions if any other _product: subcontainer is used and _product exists
-    product_object = project.packages.find_by(name: '_product') if name =~ /\A_product:\w[-+\w\.]*\z/
-    product_object || self
+    return self unless belongs_to_product?
+    project.packages.product.first
```
Comment (adrianschroeter) [inline src/api/app/models/package.rb]: this would return a product object instead of a package object, right?

project.packages.product.first.package

Comment (bgeuken) [reply src/api/app/models/package.rb]: Oh, right. But in this case we want the package that contains a product, not the product object.

I am going to rename the scope to avoid confusion.

**Comment by adrianschroeter** [comment]
On Freitag, 24. August 2018, 14:40:48 CEST wrote Björn Geuken:
> bgeuken commented on this pull request.
> 
> 
> 
> > @@ -357,8 +359,12 @@ def kiwi_image_outdated?
>  
>    def master_product_object
>      # test _product permissions if any other _product: subcontainer is used and _product exists
> -    product_object = project.packages.find_by(name: '_product') if name =~ /\A_product:\w[-+\w\.]*\z/
> -    product_object || self
> +    return self unless belongs_to_product?
> +    project.packages.product.first
> 
> Oh, right. But in this case we want the package that contains a product, not the product object.

but the old code returned a package object

> 
> I am going to rename the scope to avoid confusion.
> 
> 


-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by bgeuken** [comment]
> but the old code returned a package object

The current code as well. It's a scope on the packages relation. I just picked a bad name for it^^ Changed it now
